### DISCUSSION TITLE
How to return data from function using supabase_py?

### DISCUSSION THREAD

**Discussion by erickalfaro** [supabase/supabase#1464] [Questions] [answered]
Newbie question - how would I go about querying my function that returns an 'ID' and a 'NAME'? 

```
data = supabase.table("test_return").select("id,name").execute()
print(data)

{'data': {'hint': None, 'details': None, 'code': '42P01', 'message': 'relation "public.test_return" does not exist'}, 'status_code': 404}
```

**Comment by AngeloAnolin** [comment]
The error message seems to state that entity `"public.test_return"` does not exist.

Also, can you try the following:
```
data = supabase.table("valid_table").select("id", "name").execute()
print(data)
```

My environment is not setup at the moment but let me know if that works.

**Reply by erickalfaro** [reply]
On the supabase server side I created a custom function called test_return and am able to successfully run it and return data from the function using:

```
Select *
From test_return()
```
However I am not sure how to execute a function using supabase_py library. 

As you can see running it server side you the test_return() function has parenthesis but I'm assuming this is not how one would execute a function using the python library.

**Reply by AngeloAnolin** [reply]
I am trying to run a simple query but the console just keeps throwing an error on the package supabase-py.

Seems [supabase-py](https://github.com/supabase/supabase-py) is still in-progress.

Even the supabase [client libraries page](https://supabase.io/docs/guides/client-libraries) says that python is still WIP.

**Comment by silentworks** [comment]
Is this `test_return` a table name or a stored procedure? if it's a store procedure then you need to call it with 

```py
supabase.rpc('test_return')
```

**Reply by erickalfaro** [reply]
`test_return` is a postgres function that returns a table. 

However I am not sure how to read data from this function using supabase-py.

I tried the rpc command but it does not return anything. It simply executes the function on the server side. 

```
create or replace FUNCTION test_return()
  returns TABLE (id bigint, name text) 
AS
$func$
  SELECT id, name
  FROM countries;
$func$ 
LANGUAGE sql;
```

**Comment by ocupe** [comment]
Old but maybe helpful to others. This post explains how to return JSON from a Supabase function.
https://www.jeffreyknox.dev/blog/postgresql-functions-in-supabase/

TL;DR
```SQL
SELECT array_to_json(array_agg(row_to_json(t)))
FROM (
  SELECT
    distinct(EXTRACT(YEAR FROM created_at)) AS year,
    array_agg(distinct(EXTRACT(MONTH FROM created_at))) AS month_numbers
  FROM public.netlify_deploy_data
  GROUP BY year
  ORDER BY year ASC
)t;
```

**Reply by steve-chavez** [reply]
Nowadays you can use `json_agg(t)` instead of `array_to_json(array_agg(row_to_json(t)))`.

**Comment by J0** [comment, marked as answer]
Hey all,

Thanks for your patience!  The Python library is quite a bit further along than it was in 2021 and the query: `data = supabase.table("valid_table").select("id", "name").execute()` should work now.

Please feel free to let us know if you face any further issues!

Joel
